src/main/java/build/buildfarm/metrics/cas/CASAccessMetricsRecorder.java

```diff
@@ -0,0 +1,234 @@
+package build.buildfarm.metrics.cas;
+
+import static java.lang.String.format;
+
+import build.bazel.remote.execution.v2.Digest;
+import build.buildfarm.backplane.Backplane;
+import build.buildfarm.common.DigestUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Supplier;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.logging.Level;
+import lombok.extern.java.Log;
+
+/**
+ * This class records read and write metrics, updating the information to central storage at regular
+ * intervals. The collected data can be utilized to determine the read count of any CAS entry, query
+ * the top "n%" of frequently accessed CAS entries, and rank CAS entries based on their read
+ * frequency, among other uses.
+ */
+@Log
+public final class CASAccessMetricsRecorder {
+  private final ScheduledExecutorService scheduler;
+
+  private final Deque<Map<Digest, AtomicInteger>> readIntervalCountQueue;
+
+  private volatile Map<Digest, AtomicInteger> currentIntervalReadCount;
+
+  private final Duration casEntryReadCountWindow;
+
+  private final Duration casEntryReadCountUpdateInterval;
+
+  private final int queueLength;
+
+  private boolean running = false;
+
+  private final Backplane backplane;
+
+  final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+  int intervalNumber = 0; // Assists in unit test.
```

Comment:
I really don't like that we have this counter updated in production that we're only using for unit tests. If you can manage to move this out and keep track of the counters in different ways (Test subclassing with override for the interval operation perhaps), that would be preferred.